File: 프로그래머스/3/138475. 억억단을 외우자/억억단을 외우자.cpp
#include <string>
#include <vector>

using namespace std;
int cnt[5000001],ans[5000001];
vector<int> solution(int e, vector<int> starts) {
    vector<int> answer;
    for(int i=1;i<=e;i++){
        for(int j=i;j<=e;j+=i){
            cnt[j]++;
        }
    }
    int ma=0,idx=e;
    for(int i=e;i>=1;i--){
        if(ma<=cnt[i]){
            ma=cnt[i];
            idx=i;
        }
        ans[i]=idx;
    }
    for(int start:starts){
        answer.push_back(ans[start]);
    }
    return answer;
}

File: 프로그래머스/3/138475. 억억단을 외우자/억억단을 외우자.java
class Solution {
    
    int[] cnt = new int[5000001];
    int[] ans = new int[5000001];
    
    public int[] solution(int e, int[] starts) {
        for (int i = 1; i <= e; i++) {
            for (int j = i; j <= e; j += i) {
                cnt[j]++;
            }
        }
        
        int max = 0, idx = e;
        for (int i = e; i >= 1; i--) {
            if (max <= cnt[i]) {
                max = cnt[i];
                idx = i;
            }
            ans[i] = idx;
        }
        
        int[] answer = new int[starts.length];
        for (int i = 0; i < starts.length; i++) {
            answer[i] = ans[starts[i]];
        }
        return answer;
    }
}